### PR TITLE
Implement real validation split for TF.

### PR DESCRIPTION
**Pull request by gcp**
Split the dataset into training and validation parts to get better
information about overfitting. Small fixes to timing.

### CODE REVIEW

File: training/tf/parse.py [hunk 1]
```diff
@@ -260,8 +261,11 @@ def benchmark(parser):
         end = time.time()
         print("{} pos/sec {} secs".format( 10000. / (end - start), (end - start)))
 
-def main(args):
+def split_chunks(chunks, test_ratio):
+    splitpoint = 1 + int(len(chunks) * (1.0 - test_ratio))
+    return (chunks[:splitpoint], chunks[splitpoint:])
```
Comment (killerducky) [inline training/tf/parse.py]: Maybe do random.shuffle(chunks) first, to help the case where early chunks have weak games and later ones have strong games?

Comment (gcp) [reply training/tf/parse.py]: Yes, that's a good point.

**Comment by gcp** [comment]
Haven't tested the SL changes yet but you get the idea.

File: training/tf/parse.py [hunk 2]
```diff
@@ -270,21 +274,36 @@ def main(args):
...
     dataset = dataset.map(_parse_function)
     dataset = dataset.batch(BATCH_SIZE)
     dataset = dataset.prefetch(4)
     iterator = dataset.make_one_shot_iterator()
-    next_batch = iterator.get_next()
+    next_train_batch = iterator.get_next()
+
+    test_parser = ChunkParser(test, 1)
+    dataset = tf.data.Dataset.from_generator(
+        test_parser.parse_chunk, output_types=(tf.string))
```
Comment (gcp) [inline training/tf/parse.py]: This does need a shuffle buffer too. Without shuffling, testing 800 batches of 512 positions means we test about 1024 different games, which means the errors on the MSE measurement are very correlated.

**Comment by killerducky** [comment]
How about also in the code parsing the chunks, randomly sample a percentage (10%? 1%?) of the positions. Then we don't have to rely so heavily on having a huge dataset.shuffle. I'm afraid Ttl's experiment in the mega-thread shows we have too many positions from the same game even in the 1<<18 size shuffle.

File: training/tf/parse.py (hunk 2 above)
Comment (gcp) [reply training/tf/parse.py]: Actually, a conditional skip would be much handier here than a shuffle buffer to fix that, and more memory friendly.

**Comment by gcp** [comment]
>How about also in the code parsing the chunks, randomly sample a percentage (10%? 1%?) of the positions.

Yeah I'm thinking along the same lines. If you sample too little you risk getting CPU limited again though.

**Comment by gcp** [comment]
>we have too many positions from the same game even in the 1<<18 size shuffle.

Did that account for rotation?

**Comment by killerducky** [comment]
I don't think it accounted for rotation but presumably the 3x3 filters have already learned all rotations so the values could still be highly correlated.

**Comment by killerducky** [comment]
Link to @Ttl's results:
https://github.com/gcp/leela-zero/issues/591#issuecomment-359908639

**Comment by roy7** [comment]
Forgive me for an overly simple question, but if the training process itself is pretty slow, can't we just fetch games as needed from the chunk files and not even need to shuffle buffer them in ram? Then we can randomly pull from the whole available data without any concerns about over concentration.

I just don't know how many games need loaded or how often, but given the time it takes for 1000 steps to finish, I wasn't sure it'd matter if we did a bunch of reads off my SSD between steps.

**Comment by gcp** [comment]
>can't we just fetch games as needed from the chunk files and not even need to shuffle buffer them in ram? Then we can randomly pull from the whole available data without any concerns about over concentration.

A 250k window is about ~232G uncompressed, so the data needs to say compressed per chunk. This makes picking a single position out of entire chunk very inefficient, though.

**Comment by gcp** [comment]
>I don't think it accounted for rotation but presumably the 3x3 filters have already learned all rotations so the values could still be highly correlated.

I don't understand what you're saying here. The 3x3 filters can't rotate by themselves. They detect features in a specific orientation. Even on higher level features, they can only see a spatial distance of 1 from the previous layer, so they do need to "learn" the rotation too.

**Comment by killerducky** [comment]
That network as a whole could have constructed similar 3x3 filters for all rotations. So for feature A, it could have 8 similar 3x3 filters to detect all rotations of that feature.

Ignoring the details of how the network does it, I think it's safe to say for any position, the value head output will be highly correlated for all 8 rotations of that position. And positions from the same game are also highly correlated even if their positions are fed in with random rotations.

**Comment by gcp** [comment]
>I think it's safe to say for any position, the value head output will be highly correlated for all 8 rotations of that position.

I mean that's the goal of the training so it would be rather sad if it didn't. It doesn't actually support the argument, though.

>And positions from the same game are also highly correlated even if their positions are fed in with random rotations.

Highly correlated *where*? The residual stack has strong spatial locality, so it can't. It's probably only really true for the heads.

(This argument made me find a bug!)

**Comment by killerducky** [comment]
Yes I mean the final value output will be highly correlated for the same positions for all 8 rotations. 

Anyways with the validation set we will be able to find out how much these things matter. Enough talk for now, I leave it to you who is doing the work on this. :)

**Comment by roy7** [comment]
> (This argument made me find a bug!)

This is the great and terrible thing about the NN"s ability to generalize. It can function in spite of bugs... so we don't realize easily that a bug even exists. :)

**Comment by Ttl** [comment]
The earlier problems with value head overfitting might be related to insufficient shuffling. The correlation inside the mini-batch is probably not a problem, since it is quite small. However I think the bigger problem is correlation between the mini-batches. Previously the workers dumped all the positions from chunk to shuffle buffer in order. Big shuffle buffer helps with correlation inside the mini-batch, but even with a big shuffle buffer there is significant correlation between the positions in consecutive mini-batches since they share many of the same games. Since games were dumped in order all the positions from the same games are being trained very closely to each other and the value head outputs are highly correlated overfitting it to those games. Skipping randomly up to 8 positions helps with the problem.

**Comment by gcp** [comment]
>However I think the bigger problem is correlation between the mini-batches.

Yes, I came to the same realization. The shuffle buffer holds about 655 different games. Every mini-batch expels 512 positions from it, causing only...7 (# feeder threads!) new games to enter.

>Skipping randomly up to 8 positions helps with the problem.

It's even a bit more effective than this. It doesn't just skip positions, it ends up taking a randomly ordered subsample out of the *entire* chunk. Given that the chunks contains 32 games, this means each new batch pulls from about 224 different games.

I had some weird performance effects from this, so I can't quantify yet what drop factor we can use without losing speed.

**Comment by jkiliani** [comment]
Would it make sense to make a background process for the training machine that pre-shuffles chunks from the same dataset (i.e. produced by self-play from the same network)? At the moment, the training chunks are completely ordered, does it serve any purpose to keep it that way? If the chunks were not ordered anymore at the time of parsing, maybe you could use a smaller shuffle buffer...

**Comment by gcp** [comment]
>Would it make sense to make a background process for the training machine that pre-shuffles chunks from the same dataset (i.e. produced by self-play from the same network)?

I'm not sure what a background process would gain you?

**Comment by gcp** [comment]
One problem I've got is:

```
        self.next_batch = \
            tf.cond(self.training,
                    lambda: next_train_batch, lambda: next_test_batch)
```

Because of how TensorFlow works, both sides of that branch are evaluated. So I'm burning twice as much CPU as intended.

**Comment by jkiliani** [comment]
> I'm not sure what a background process would gain you? Like, what's the difference with what we're doing now?

Not sure if a background process is the best way to do this, but I'm asking why not read in a number of chunks, shuffle the training data in them both internally and between the chunks, and write them back out again into new, shuffled chunks. That way, any time parsing uses these chunks in the future (quite a number of times, while it remains inside the training window), it would be starting from chunks that are already shuffled, instead of completely ordered ones.

**Comment by gcp** [comment]
>Not sure if a background process is the best way to do this, but I'm asking why not read in a number of chunks, shuffle the training data in them both internally and between the chunks, and write them back out again into new, shuffled chunks.

That's what it's doing now on the fly, except that it doesn't write out the result to disk, which would be slow, and require a pre-processing pass (else the initial learning steps would still have the same problem) which delays things?

>That way, any time parsing uses these chunks in the future (quite a number of times, while it remains inside the training window)

There's 250k * ~400 = 100M positions in the window. 128k steps * 512 positions = 65M positions used.

**Comment by gcp** [comment]
>I realised you'd be giving up the ability to easily filter out particular games from the training data with this, so I guess it depends on whether you have concerns about corrupted data at the time...

I need to take in the data from the original DB (mongo) and write out the chunk files anyhow, which is the point the filtering checks the IP of the source data. But shuffling there takes the same resources as shuffling later on.

**Comment by gcp** [comment]
@ywrt If you're still around I could use your help right about now :)

**Comment by killerducky** [comment]
I guess you already maxed your RAM usage? If not you could double dataset.shuffle a few more times.

**Comment by gcp** [comment]
>I guess you already maxed your RAM usage?

Yeah basically. Hard to fix this with just a shuffle buffer. If I could work out a way to avoid the tf.cond silliness the solution here would be OK.

**Comment by marcocalignano** [comment]
I have a question why just take one random simmentry and not all 8 of them? You could generate more data and distribute them overr othar batches. [L160](https://github.com/gcp/leela-zero/blob/23474b6bc97411752d44058ed605ecbddc17b949/training/tf/parse.py#L160)

**Comment by gcp** [comment]
>and distribute them overr othar batches.

Yes but the question is how. That's exactly the problem here.

**Comment by ywrt** [comment]
@gcp I'm back. I skimmed this thread. Could you maybe summarize what you think the problem is?

My guess after skimming the thread is that the mini-batches are much too internally correlated, and you're looking to have something closer to a random sample over all the moves from all the games in the window?

As an aside: Do you have a pointer to the code you're using to pull from mongo db? (maybe add it to the training directory somewhere?)

One of the challenges I had before is that the format that moves are stored in makes it much harder to run fast.

**Comment by gcp** [comment]
> Could you maybe summarize what you think the problem is?

https://github.com/gcp/leela-zero/pull/747#issuecomment-360871125

I'm open to other ideas to make the shuffling better and fast, but this is currently bothering me. I'm trying to find a clean way to pull from the right dataset, without having TF run both parsers at once.

>As an aside: Do you have a pointer to the code you're using to pull from mongo db? (maybe add it to the training directory somewhere?)

https://pastebin.mozilla.org/9076742

**Comment by marcocalignano** [comment]
> Yes but the question is how.

something like this?

```
    def task(self, chunks, workers):
        while True:
            random.shuffle(chunks)
            for chunk in chunks:
                with gzip.open(chunk, 'r') as chunk_file:
                    file_content = chunk_file.readlines()
                    item_count = len(file_content) // DATA_ITEM_LINES
                    for item_idx in range(item_count):
                        pick_offset = item_idx * DATA_ITEM_LINES
                        item = file_content[pick_offset:pick_offset + DATA_ITEM_LINES]
                        str_items = [str(line, 'ascii') for line in item]
                        symmetry = 0
                        for write in writes:
                            success, data = self.convert_train_data(str_items, symmetry)
                            if success:
                                # Send it down the pipe.
                                writer.send_bytes(data)
                            symmetry = (symmetry + 1) % 8

```

**Comment by gcp** [comment]
>something like this?

That doesn't distribute them over batches. The inability of the shuffle buffer to do so amounts to most of the discussion in this thread.

**Comment by ywrt** [comment]
(After a closer read)

I think you'd find this easier if the partitioning was done on the mongodb export side?

My first though would be to hash-partition the moves into separate chunk sets, and then use a feedable iterator to switch between the training and validation datasets.

So the export side, something like:
```
  for m in moves:
    if hash(m) < MAX_HASH_VALUE * 0.01:  # 1% validation sample.
      write_to_chunk_with_prefix("validate", m)
    else:
      write_to_chunk_with_prefix("train", m)
```

and then on the TF side, build two separate Datasets and use a placeholder to feed them separately into training and validation.

If you like, I can sketch out a patch for this on the TF side?

Re shuffling: I agree that the limited shuffling is a real concern. It might be worth doing a random shuffle on export? 

This is really a RAM consideration: Ideally, we'd have `number_of_partitions = available_ram / (2* 1024 * 1024)`, which spreads as wide as possible while still being able to write to disk in 2MB chunks (so we stay throughput bound rather than spindle bound).

Which translates to: maintain N buffers in RAM, add each move to a random buffer. When a buffer reaches ~2MB, flush it to a random chunk file that's not yet full. 

The next step is to do the reverse on the TF side: read from K chunks in parallel, reading into 2MB buffers (to stay disk efficient), and merge the chunks to produce the final dataflow in TF.
This would reduce in a not-completely-terrible approximation to uniform sampling, without overly much CPU overhead. (Obv, we're trading off increased RAM to reduce the CPU cost of the shuffle).

Depends on how crazy you want to get :)

**Comment by isty2e** [comment]
My programming skill is somewhere between abysmal and mediocre, so this might be a silly question. If I were to do this, I probably would separated the dataset first and let the train and test dataset to have different iterators, which also probably could bypass the double evaluation of tf.cond()... I suppose. Is there a reason the split is not implemented this way?

**Comment by marcocalignano** [comment]
@gcp But maybe you can create pipes between threads and each threads write a different symmetry on a pipe from a different thread. Could that work?

**Comment by gcp** [comment]
> If I were to do this, I probably would separated the dataset first and let the train and test dataset to have different iterators, which also probably could bypass the double evaluation of tf.cond()

This is exactly how it's implemented.

**Comment by killerducky** [comment]
Is it true that the main reason the chunks are written as contiguous games is for the bad game filtering process? How about the bad game filter process is run completely separately one time only, and each game is marked good/bad in the database. Then when reading out from the database and creating the chunks they can be written out pre-randomized.

 BTW are there still bad games now that the OpenCL bug is detected automatically?

**Comment by gcp** [comment]
>build two separate Datasets and use a placeholder to feed them separately into training and validation.

That's what the code does. Look in this pull request.

**Comment by gcp** [comment]
>Then when reading out from the database and creating the chunks they can be written out pre-randomized.

As explained previously here, this makes no difference to the computing time required, and it ruins the compression.

**Comment by ywrt** [comment]
Sorry, I was unclear. The placeholder should be for the iterator. Something like:
In main:
```
  train_handle = sess.run(next_train_batch.string_handle())
  valid_handle = sess.run(next_test_batch.string_handle())

  handle = tf.placeholder(tf.string, shape=[])
  iterator = tf.data.Iterator.from_string_handle(
      handle, dataset.output_types, dataset.output_shapes)
  next_batch = iterator.get_next()

  TFProcess(next_batch, train_handle, valid_handle)
```

and in TFProcess for training:
```
  # Run training for this batch
        policy_loss, mse_loss, reg_term, _, _ = self.session.run(
            [self.policy_loss, self.mse_loss, self.reg_term, self.train_op,
                self.next_batch],
                feed_dict={self.training: True, handle: train_handle})
```
and then validation
```
         for _ in range(0, 10):
                train_accuracy, train_mse, _ = self.session.run(
                    [self.accuracy, self.mse_loss, self.next_batch],
                    feed_dict={self.training: False, handle: valid_handle})
```

Is that clearer? As I mentioned, happy to sketch out a patch for this if you want.

**Comment by gcp** [comment]
Thanks, that gives me some idea how to solve it.

**Comment by ywrt** [comment]
Re the shuffle: If you want something dumb and easy, changing the workers as follows would help reduce intra-batch correlation.
```
    CHUNKS_TO_MERGE=8
    def chunk_convert(self, chunk):
        with gzip.open(chunk, 'r') as chunk_file:
            file_content = chunk_file.readlines()
            item_count = len(file_content) // DATA_ITEM_LINES
            for item_idx in range(item_count):
                pick_offset = item_idx * DATA_ITEM_LINES
                item = file_content[pick_offset:pick_offset + DATA_ITEM_LINES]
                str_items = [str(line, 'ascii') for line in item]
                # Pick a random symmetry to apply
                symmetry = random.randrange(8)
                success, data = self.convert_train_data(str_items, symmetry)
                if success:
                    yield data

    def task(self, chunks, writer):
        while True:
            random.shuffle(chunks)
            chunk_list = chunks
            while chunk_list:
                c = chunk_list[:CHUNKS_TO_MERGE]
                chunk_list = chunk_list[CHUNKS_TO_MERGE:]
                iters = [ self.chunk_convert(i) for i in c]
                for m in zip(iters):
                    writer.send_bytes(m)

```

This is pretty dumb as it needlessly pulls the entire chunk file into RAM.

Note also, that the shuffle buffer in the Dataset pipeline was deliberately Dataset.shuffle().map(), not Dataset.map().shuffle(), so that the shuffle operates on packed strings.

If you want a larger shuffle buffer, right now the ChunkParser assumes that the 'probs' can be arbitrary floats, when in reality it's a one-hot vector (all zero except a single one), so the string representing the move can be shrunk to ~ 25% of it's current size, similarly making the shuffle buffer a quarter of it's current size (it would also make both the parsing faster and the TF graph faster but that's by-the-by).

**Comment by gcp** [comment]
>right now the ChunkParser assumes that the 'probs' can be arbitrary floats, when in reality it's a one-hot vector (all zero except a single one)

That's only true for supervised learning data. RL data has a true probability map.

**Comment by gcp** [comment]
Ok together with your hint and https://www.tensorflow.org/programmers_guide/datasets#creating_an_iterator I think I got something.

File: training/tf/parse.py [hunk 3]
```diff
@@ -270,21 +277,34 @@ def main(args):
...
     dataset = dataset.batch(BATCH_SIZE)
     dataset = dataset.prefetch(4)
-    iterator = dataset.make_one_shot_iterator()
-    next_batch = iterator.get_next()
+    train_iterator = dataset.make_one_shot_iterator()
+
+    test_parser = ChunkParser(test, 1)
+    dataset = tf.data.Dataset.from_generator(
+        test_parser.parse_chunk, output_types=(tf.string))
+    dataset = dataset.map(_parse_function)
```
Comment (ywrt) [inline training/tf/parse.py]: Might be a good idea to shuffle the validation dataset too?

Comment (gcp) [reply training/tf/parse.py]: Not sure, didn't we just discover the shuffle buffer may not help all that much unless its made much larger?

**Comment by marcocalignano** [comment]
@gcp and if you do all the task work in one task and move only `self.convert_train_data(str_items, symmetry)` in a different thread you could distribute the data already shuffle and with a different symmetry to each task.

**Comment by gcp** [comment]
>@gcp and if you do all the task work in one task

Then it'll bottleneck even harder in the ungzip+skip, which is that actual bottleneck...
